lib/internal/modules/run_main.js

```diff
@@ -119,7 +119,19 @@ function executeUserEntryPoint(main = process.argv[1]) {
   } else {
     // Module._load is the monkey-patchable CJS module loader.
     const { Module } = require('internal/modules/cjs/loader');
-    Module._load(main, null, true);
+    try {
+      Module._load(main, null, true);
+    } catch (err) {
+      if (getOptionValue('--experimental-detect-module')) {
+        const { isModuleSyntaxError } = require('internal/modules/helpers');
+        if (isModuleSyntaxError(err)) {
+          // If the error is a module syntax error, we should use the ESM loader instead.
+          runMainESM(resolvedMain || main);
+          return;
+        }
+      }
+      throw err;
```

Comment:
I think this might be why the runtime error source lines get swallowed. Maybe can move the detection deeper inside `wrapSafe` so that only `SyntaxError` gets caught, and since we already take care of the source lines in the compilation APIs, they can be preserved. (We don't do that for other runtime errors that come out of the actual loading/invocation of the module)